// unplugin-preprocessor-directives/dist/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});require('./chunk-ZBPRDZS4.cjs');



















var _chunkW7WZVURScjs = require('./chunk-W7WZVURS.cjs');



















exports.Context = _chunkW7WZVURScjs.Context; exports.Lexer = _chunkW7WZVURScjs.Lexer; exports.MessageDirective = _chunkW7WZVURScjs.MessageDirective; exports.Parser = _chunkW7WZVURScjs.Parser; exports.comments = _chunkW7WZVURScjs.comments; exports.createProgramNode = _chunkW7WZVURScjs.createProgramNode; exports.defineDirective = _chunkW7WZVURScjs.defineDirective; exports.findComment = _chunkW7WZVURScjs.findComment; exports.ifDirective = _chunkW7WZVURScjs.ifDirective; exports.isComment = _chunkW7WZVURScjs.isComment; exports.parseComment = _chunkW7WZVURScjs.parseComment; exports.resolveConditional = _chunkW7WZVURScjs.resolveConditional; exports.resolveOptions = _chunkW7WZVURScjs.resolveOptions; exports.simpleMatchToken = _chunkW7WZVURScjs.simpleMatchToken; exports.sortUserDirectives = _chunkW7WZVURScjs.sortUserDirectives; exports.theDefineDirective = _chunkW7WZVURScjs.theDefineDirective; exports.unplugin = _chunkW7WZVURScjs.unplugin; exports.unpluginFactory = _chunkW7WZVURScjs.unpluginFactory;
exports.default = module.exports;
